### PR TITLE
[SQL] Fix incorrect simplification of constant casts to integers

### LINKED ISSUE #1180: [SQL] InternalCompilerError:  Expected function to return Tuple1<i32> but it returns Tuple1<i32?>
Creating a test as such, causes an internal compiler error. The compiler expects the function to return `Tuple1<i32>` but gets `Tuple1<i32?>` which is not really a valid Rust type. 

```java
    @Test
    public void ICE() {
        this.q(
                "SELECT '1_000'::INT4;\n" +
                        " int2 \n" +
                        "------\n" +
                        " 1000"
        );
    }
```
Even if the SQL compiler doesn't support underscores in numbers, it shouldn't cause an InternalCompilerError. 

Stack trace:
```java
org.dbsp.sqlCompiler.compiler.errors.InternalCompilerError: Map 146: Expected function to return Tuple1<i32> but it returns Tuple1<i32?>

	at org.dbsp.sqlCompiler.circuit.operator.DBSPOperator.checkResultType(DBSPOperator.java:126)
	at org.dbsp.sqlCompiler.circuit.operator.DBSPMapOperator.<init>(DBSPMapOperator.java:44)
	at org.dbsp.sqlCompiler.compiler.visitors.outer.CircuitRewriter.postorder(CircuitRewriter.java:278)
	at org.dbsp.sqlCompiler.circuit.operator.DBSPMapOperator.accept(DBSPMapOperator.java:53)
	at org.dbsp.sqlCompiler.compiler.visitors.outer.CircuitRewriter.preorder(CircuitRewriter.java:286)
	at org.dbsp.sqlCompiler.circuit.DBSPPartialCircuit.accept(DBSPPartialCircuit.java:109)
	at org.dbsp.sqlCompiler.circuit.DBSPCircuit.accept(DBSPCircuit.java:64)
	at org.dbsp.sqlCompiler.compiler.visitors.outer.CircuitCloneVisitor.apply(CircuitCloneVisitor.java:342)
	at org.dbsp.sqlCompiler.compiler.visitors.outer.CircuitCloneVisitor.apply(CircuitCloneVisitor.java:48)
	at org.dbsp.sqlCompiler.compiler.visitors.outer.Passes.apply(Passes.java:76)
	at org.dbsp.sqlCompiler.compiler.visitors.outer.Passes.apply(Passes.java:36)
	at org.dbsp.sqlCompiler.compiler.visitors.outer.CircuitOptimizer.optimize(CircuitOptimizer.java:118)
	at org.dbsp.sqlCompiler.compiler.DBSPCompiler.optimize(DBSPCompiler.java:330)
	at org.dbsp.sqlCompiler.compiler.sql.SqlIoTest.compare(SqlIoTest.java:532)
	at org.dbsp.sqlCompiler.compiler.sql.SqlIoTest.q(SqlIoTest.java:555)
	at org.dbsp.sqlCompiler.compiler.sql.SqlIoTest.q(SqlIoTest.java:561)
	at org.dbsp.sqlCompiler.compiler.sql.postgres.PostgresInt2Tests.ICE(PostgresInt2Tests.java:400)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:232)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:55)


Process finished with exit code 255

```

### PR DESCRIPTION
**Pull request by mihaibudiu**
Is this a user-visible change (yes/no): no

Fixes #1180
Also, made the documentation on casts more explicit for abnormal behaviors.

### CODE REVIEW

**Comment by mihaibudiu** [comment]
The expression simplification code mistakenly assumed that `CAST("foo" AS INTEGER)` returns `NULL`. In fact it should return `0`.
The sad thing is that this kind of evaluation is baked in many times in the tool chain, several times in Calcite, and also in our compiler and runtime.

**Review by abhizer — APPROVED** [review]

File: docs/sql/casts.md [hunk 1]
```diff
@@ -27,4 +27,10 @@ range values.
 Conversions from decimal and floating point types to integer types
 always truncate the decimal digits (round towards zero).  For example,
 `CAST(2.9 AS INTEGER)` returns 2, while `CAST(-2.9 AS INTEGER)`
-returns -2.
\ No newline at end of file
+returns -2.
+
+Casts of strings to numeric types produce the result zero (0) when the
```
Comment (lalithsuresh) [inline docs/sql/casts.md:32]: If this is standard SQL or so, give a reference.

**Review by lalithsuresh — APPROVED** [review]

File: docs/sql/casts.md (hunk 1 above)
Comment (mihaibudiu) [reply docs/sql/casts.md:32]: You are funny. The sql standards are all paywalled. And no one implemented the standard anyway.

Comment (lalithsuresh) [reply docs/sql/casts.md:32]: Are other DBs publicly documented to behave this way? Postgres or MySQL for example?

Comment (mihaibudiu) [reply docs/sql/casts.md:32]: In fact the standard says clearly that in this case one should trigger a runtime error. Not null, not zero.
I will check if it's a Calcite bug or a "feature".

Comment (mihaibudiu) [reply docs/sql/casts.md:32]: I will change the documentation for this case and file an issue.

**Comment by mihaibudiu** [comment]
As @lalithsuresh correctly noticed, returning 0 for a cast of an invalid string is wrong. So I have implemented the behavior which fails at runtime. (And gives a warning at compile-time if this is detected.)

This commit also adds a function `runtimeFail` for writing tests that are expected to fail at runtime.

**Comment by mihaibudiu** [comment]
This also adds utility functions `reportWarning` and `reportError`. I always forgot which one has a `true` flag. Now I no longer have to think.

**Comment by mihaibudiu** [comment]
@abhizer we will have to write a bunch of tests for the cases described in the documentation. In particular, we'll need some negative tests for numbers including invalid strings and NULLs.

**Review by lalithsuresh — APPROVED** [review]

**Comment by mihaibudiu** [comment]
Fixes #1190

File: sql-to-dbsp-compiler/lib/sqllib/src/casts.rs [hunk 2]
```diff
@@ -258,21 +258,16 @@ pub fn cast_to_decimal_fN(value: Option<F32>, precision: u32, scale: i32) -> Dec
 
 #[inline]
 pub fn cast_to_decimal_s(value: String, precision: u32, scale: i32) -> Decimal {
-    let result = match value.trim().parse().ok() {
-        None => Decimal::zero(),
-        Some(x) => x,
-    };
+    let result = value.trim().parse().unwrap();
+    println!("{}", result);
```
Comment (abhizer) [inline sql-to-dbsp-compiler/lib/sqllib/src/casts.rs:262]: We should probably remove this print.

Comment (mihaibudiu) [reply sql-to-dbsp-compiler/lib/sqllib/src/casts.rs:262]: Yes

**Pull request merged**